File: src/main/java/com/ptah/camunda/service/CamundaService.java
package com.ptah.camunda.service;

import com.ptah.camunda.model.dto.Activity;
import com.ptah.camunda.model.dto.Process;
import org.camunda.bpm.engine.ProcessEngine;
import org.camunda.bpm.engine.RuntimeService;
import org.camunda.bpm.engine.runtime.ActivityInstance;
import org.camunda.bpm.engine.runtime.ProcessInstance;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.Arrays;
import java.util.List;
import java.util.Map;

@Service
public class CamundaService {

    @Autowired
    private ProcessEngine processEngine;

    @Autowired
    private RuntimeService runtimeService;

    public Process startWorkflow(String processDefinitionKey, Map<String, Object> variables) {
        String processInstanceId = runtimeService.startProcessInstanceByKey(processDefinitionKey, variables).getProcessInstanceId();
        return getProcessInstance(processInstanceId);
    }

    public Process getProcessInstance(String processInstanceId) {
        ProcessInstance processInstance = runtimeService.createProcessInstanceQuery().processInstanceId(processInstanceId).singleResult();
        if (processInstance != null) {
            ActivityInstance activityInstance = runtimeService.getActivityInstance(processInstanceId);
            List<Activity> activities = Arrays.stream(activityInstance.getChildActivityInstances()).map(childActivityInstance -> new Activity(childActivityInstance.getActivityName(), childActivityInstance.getActivityType())).toList();
            return Process.builder().activeActivities(activities).processInstanceId(processInstanceId).businessKey(processInstance.getBusinessKey()).variables(getVariables(processInstanceId)).build();
        }
        throw new RuntimeException();
    }

    public Map<String, Object> getVariables(String executionId) {
        return runtimeService.getVariables(executionId);
    }

    public Process sendMessage(String messageName, String processInstanceId, Map<String, Object> params) {
        runtimeService.createMessageCorrelation(messageName).processInstanceId(processInstanceId).setVariables(params).correlate();
        return getProcessInstance(processInstanceId);
    }
}

File: backend/src/main/kotlin/com/github/jarryzhou/ptah/camunda/service/CamundaService.kt
package com.github.jarryzhou.ptah.camunda.service

import com.ptah.camunda.model.dto.Activity
import com.ptah.camunda.model.dto.Process
import org.camunda.bpm.engine.ProcessEngine
import org.camunda.bpm.engine.RuntimeService
import org.springframework.stereotype.Service

@Service
class CamundaService(
    private var processEngine: ProcessEngine,
    private var runtimeService: RuntimeService
) {
    fun startWorkflow(processDefinitionKey: String?, variables: Map<String, Any?>?): Process {
        val processInstanceId =
            runtimeService.startProcessInstanceByKey(processDefinitionKey, variables).processInstanceId
        return getProcessInstance(processInstanceId)
    }

    fun getProcessInstance(processInstanceId: String?): Process {
        val processInstance =
            runtimeService.createProcessInstanceQuery().processInstanceId(processInstanceId).singleResult()
        if (processInstance != null) {
            val activityInstance = runtimeService.getActivityInstance(processInstanceId)
            val activities = activityInstance.childActivityInstances.map {
                Activity(
                    name = it.activityName,
                    type = it.activityType
                )
            }
            return Process().apply {
                activeActivities = activities
                this.processInstanceId = processInstanceId
                businessKey = processInstance.businessKey
                variables = getVariables(processInstanceId)
            }
        }
        throw RuntimeException()
    }

    fun getVariables(executionId: String?): Map<String, Any> = runtimeService.getVariables(executionId)

    fun sendMessage(messageName: String?, processInstanceId: String?, params: Map<String?, Any?>?): Process {
        runtimeService.createMessageCorrelation(messageName).processInstanceId(processInstanceId).setVariables(params)
            .correlate()
        return getProcessInstance(processInstanceId)
    }
}
